test/integration/targets/zypper/tasks/zypper.yml

```diff
@@ -286,14 +306,65 @@
       - zypper_patch.msg.startswith('Can not remove patches.')
 
 - name: try rm URL
-  zypper: name=http://download.opensuse.org/repositories/openSUSE:/Leap:/{{ ansible_distribution_version }}/standard/x86_64/hello-2.9-6.2.x86_64.rpm state=absent
+  zypper:
+    name: http://download.opensuse.org/repositories/openSUSE:/Leap:/{{ ansible_distribution_version }}/standard/x86_64/hello-2.9-6.2.x86_64.rpm
+    state: absent
   ignore_errors: yes
   register: zypper_rm
 - assert:
     that: 
       - zypper_rm is failed
       - zypper_rm.msg.startswith('Can not remove via URL.')
 
+- name: remove pattern update_test
+  zypper:
+    name: update_test
+    type: pattern
+    state: absent
+
+- name: install pattern update_test
+  zypper:
+    name: update_test
+    type: pattern
+    state: present
+  register: zypper_install_pattern1
+  
+- name: install pattern update_test again
+  zypper:
+    name: update_test
+    type: pattern
+    state: present
+  register: zypper_install_pattern2
+
+- assert:
+    that:
+      - zypper_install_pattern1 is changed
+      - zypper_install_pattern2 is not changed
+
+- name: remove hello
+  zypper:
+    name: hello
+    state: absent
+
+- name: install via URL
+  zypper:
+    state: present
+    name: http://download.opensuse.org/distribution/leap/42.3/repo/oss/suse/x86_64/hello-2.9-9.6.x86_64.rpm
```

Comment:
This should use `{{ ansible_distribution_version }}` instead of a hard-coded version. Otherwise it will fail in the future when that version is no longer in the repository.

It will also need to query to find the correct RPM name, as it changes from release to release. The only reason it's not an issue above is because the URL doesn't need to exist in order to cause the failure to remove by URL.

Reply:
The package URL specified here can be installed on opensuse version including 42.1, 42.2, 42.3 and tumbleweed. Also the URL is expected to work for around 2 years and I expect it to work with opensuse 15.x as well.

Generating a URL based on the distribution_version is more work, since the available package versions change and I guess some coding doing this will also break in a period of more than one year, so I find having a fixed URL with a stable and very compatible package the easier to understand and easier to maintain solution.

Reply:
@robinro and @mattclay.
1. There are two different urls, which should be present in the same way or even be the same url:
`http://download.opensuse.org/repositories/openSUSE:/Leap:/{{ ansible_distribution_version }}/standard/x86_64/hello-2.9-6.2.x86_64.rpm`
`http://download.opensuse.org/distribution/leap/42.3/repo/oss/suse/x86_64/hello-2.9-9.6.x86_64.rpm`
1.  IMHO, it looks more like a configuration. Hardcoding these values into different places is not good idea. URL or package version can be changed in future. It is a good approach(as configuration setup) in yum module: https://github.com/ansible/ansible/blob/64a355cb7ed3c7983758df053687c177baf795eb/test/integration/targets/yum/tasks/yum.yml#L474:L486